File: uploadservice-ftp/src/main/java/net/gotev/uploadservice/ftp/FTPUploadTaskParameters.java
package net.gotev.uploadservice.ftp;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * FTP upload parameters.
 * @author Aleksandar Gotev
 */
public class FTPUploadTaskParameters implements Parcelable {

    protected static final String PARAM_FTP_TASK_PARAMETERS = "ftpTaskParameters";

    /**
     * The default FTP connection timeout in milliseconds.
     */
    public static final int DEFAULT_CONNECT_TIMEOUT = 15000;

    /**
     * The default FTP socket timeout in milliseconds.
     */
    public static final int DEFAULT_SOCKET_TIMEOUT = 30000;

    /**
     * The default protocol to use when FTP over SSL is enabled (FTPS).
     */
    public static final String DEFAULT_SECURE_SOCKET_PROTOCOL = "TLS";

    private int port;
    private String username;
    private String password;
    private int connectTimeout;
    private int socketTimeout;
    private boolean compressedFileTransfer;
    private String createdDirectoriesPermissions;
    private boolean useSSL;
    private boolean implicitSecurity;
    private String secureSocketProtocol = DEFAULT_SECURE_SOCKET_PROTOCOL;

    public FTPUploadTaskParameters() {

    }

    // This is used to regenerate the object.
    // All Parcelables must have a CREATOR that implements these two methods
    public static final Creator<FTPUploadTaskParameters> CREATOR =
            new Creator<FTPUploadTaskParameters>() {
                @Override
                public FTPUploadTaskParameters createFromParcel(final Parcel in) {
                    return new FTPUploadTaskParameters(in);
                }

                @Override
                public FTPUploadTaskParameters[] newArray(final int size) {
                    return new FTPUploadTaskParameters[size];
                }
            };

    @Override
    public void writeToParcel(Parcel parcel, int arg1) {
        parcel.writeInt(port);
        parcel.writeString(username);
        parcel.writeString(password);
        parcel.writeInt(connectTimeout);
        parcel.writeInt(socketTimeout);
        parcel.writeByte((byte) (compressedFileTransfer ? 1 : 0));
        parcel.writeString(createdDirectoriesPermissions);
        parcel.writeByte((byte) (useSSL ? 1 : 0));
        parcel.writeByte((byte) (implicitSecurity ? 1 : 0));
        parcel.writeString(secureSocketProtocol);
    }

    private FTPUploadTaskParameters(Parcel in) {
        port = in.readInt();
        username = in.readString();
        password = in.readString();
        connectTimeout = in.readInt();
        socketTimeout = in.readInt();
        compressedFileTransfer = in.readByte() == 1;
        createdDirectoriesPermissions = in.readString();
        useSSL = in.readByte() == 1;
        implicitSecurity = in.readByte() == 1;
        secureSocketProtocol = in.readString();
    }

    @Override
    public int describeContents() {
        return 0;
    }

    public int getPort() {
        return port;
    }

    public FTPUploadTaskParameters setPort(int port) {
        this.port = port;
        return this;
    }

    public String getUsername() {
        return username;
    }

    public FTPUploadTaskParameters setUsername(String username) {
        this.username = username;
        return this;
    }

    public String getPassword() {
        return password;
    }

    public FTPUploadTaskParameters setPassword(String password) {
        this.password = password;
        return this;
    }

    public int getConnectTimeout() {
        return connectTimeout <= 0 ? DEFAULT_CONNECT_TIMEOUT : socketTimeout;
    }

    public FTPUploadTaskParameters setConnectTimeout(int connectTimeout) {
        this.connectTimeout = connectTimeout;
        return this;
    }

    public int getSocketTimeout() {
        return socketTimeout <= 0 ? DEFAULT_SOCKET_TIMEOUT : socketTimeout;
    }

    public FTPUploadTaskParameters setSocketTimeout(int socketTimeout) {
        this.socketTimeout = socketTimeout;
        return this;
    }

    public boolean isCompressedFileTransfer() {
        return compressedFileTransfer;
    }

    public FTPUploadTaskParameters setCompressedFileTransfer(boolean compressedFileTransfer) {
        this.compressedFileTransfer = compressedFileTransfer;
        return this;
    }

    public String getCreatedDirectoriesPermissions() {
        return createdDirectoriesPermissions;
    }

    public FTPUploadTaskParameters setCreatedDirectoriesPermissions(String createdDirectoriesPermissions) {
        this.createdDirectoriesPermissions = createdDirectoriesPermissions;
        return this;
    }

    public boolean isUseSSL() {
        return useSSL;
    }

    public FTPUploadTaskParameters setUseSSL(boolean useSSL) {
        this.useSSL = useSSL;
        return this;
    }

    public boolean isImplicitSecurity() {
        return implicitSecurity;
    }

    public FTPUploadTaskParameters setImplicitSecurity(boolean implicitSecurity) {
        this.implicitSecurity = implicitSecurity;
        return this;
    }

    public String getSecureSocketProtocol() {
        return secureSocketProtocol;
    }

    public FTPUploadTaskParameters setSecureSocketProtocol(String secureSocketProtocol) {
        this.secureSocketProtocol = secureSocketProtocol;
        return this;
    }
}

File: uploadservice-ftp/src/main/java/net/gotev/uploadservice/ftp/FTPUploadTaskParameters.kt
package net.gotev.uploadservice.ftp

import android.os.Parcelable
import kotlinx.parcelize.Parcelize
import net.gotev.uploadservice.persistence.Persistable
import net.gotev.uploadservice.persistence.PersistableData

/**
 * FTP upload parameters.
 * @author Aleksandar Gotev
 */
@Parcelize
data class FTPUploadTaskParameters(
    var port: Int,
    var username: String = "anonymous",
    var password: String = "",
    var connectTimeout: Int = DEFAULT_CONNECT_TIMEOUT,
    var socketTimeout: Int = DEFAULT_SOCKET_TIMEOUT,
    @get:JvmName("isCompressedFileTransfer")
    var compressedFileTransfer: Boolean = false,
    var createdDirectoriesPermissions: String? = null,
    var useSSL: Boolean = false,
    @get:JvmName("isImplicitSecurity")
    var implicitSecurity: Boolean = false,
    var secureSocketProtocol: String = DEFAULT_SECURE_SOCKET_PROTOCOL
) : Parcelable, Persistable {
    companion object : Persistable.Creator<FTPUploadTaskParameters> {
        /**
         * The default FTP connection timeout in milliseconds.
         */
        const val DEFAULT_CONNECT_TIMEOUT = 15000

        /**
         * The default FTP socket timeout in milliseconds.
         */
        const val DEFAULT_SOCKET_TIMEOUT = 30000

        /**
         * The default protocol to use when FTP over SSL is enabled (FTPS).
         */
        const val DEFAULT_SECURE_SOCKET_PROTOCOL = "TLS"

        private object CodingKeys {
            const val port = "port"
            const val username = "user"
            const val password = "pwd"
            const val connectTimeout = "cTimeout"
            const val socketTimeout = "soTimeout"
            const val compressed = "compressed"
            const val permissions = "perms"
            const val ssl = "ssl"
            const val implicit = "implicit"
            const val sslProtocol = "sslProto"
        }

        override fun createFromPersistableData(data: PersistableData) = FTPUploadTaskParameters(
            port = data.getInt(CodingKeys.port),
            username = data.getString(CodingKeys.username),
            password = data.getString(CodingKeys.password),
            connectTimeout = data.getInt(CodingKeys.connectTimeout),
            socketTimeout = data.getInt(CodingKeys.socketTimeout),
            compressedFileTransfer = data.getBoolean(CodingKeys.compressed),
            createdDirectoriesPermissions = try {
                data.getString(CodingKeys.permissions)
            } catch (exc: Throwable) {
                null
            },
            useSSL = data.getBoolean(CodingKeys.ssl),
            implicitSecurity = data.getBoolean(CodingKeys.implicit),
            secureSocketProtocol = data.getString(CodingKeys.sslProtocol)
        )
    }

    override fun toPersistableData() = PersistableData().apply {
        putInt(CodingKeys.port, port)
        putString(CodingKeys.username, username)
        putString(CodingKeys.password, password)
        putInt(CodingKeys.connectTimeout, connectTimeout)
        putInt(CodingKeys.socketTimeout, socketTimeout)
        putBoolean(CodingKeys.compressed, compressedFileTransfer)
        createdDirectoriesPermissions?.let { putString(CodingKeys.permissions, it) }
        putBoolean(CodingKeys.ssl, useSSL)
        putBoolean(CodingKeys.implicit, implicitSecurity)
        putString(CodingKeys.sslProtocol, secureSocketProtocol)
    }
}
